File: pyqt/start_panel.py
# !/usr/bin/python3
# -*- coding: utf-8 -*-
import os
import sys
import math

from PySide6.QtWidgets import QApplication, QMainWindow, QMessageBox, QFileDialog, QWidget
from PySide6.QtCore import Slot, Signal, Qt, QTimer, QPoint, QRect
from PySide6.QtGui import QPainter, QPen, QBrush

import global_config
from calibration_widget import CalibrationWidget
from calibration_result_widget import CalibrationResultWidget
from setting_dialog import SettingDialog
from image_show_widget import ImageShowWidget
from ui.ui_start_panel import Ui_start_panel


class EyePosShow(QWidget):
    def __init__(self, parent=None):
        super(EyePosShow, self).__init__(parent)
        self.p_x = [0.5, 0.5, 0.5]
        self.p_y = [0.5, 0.5, 0.5]
        self.current_rad = 10
        self.IsPainter = False

    def paintEvent(self, event):
        super().paintEvent(event)
        if self.IsPainter == True:
            self.W = self.width()
            self.H = self.height() * 0.8
            painter = QPainter()
            painter.begin(self)
            pen = QPen(Qt.white, 1, Qt.SolidLine)
            painter.setPen(pen)
            brush = QBrush()
            brush.setColor(Qt.white)
            brush.setStyle(Qt.SolidPattern)
            painter.setBrush(brush)
            rect1 = QRect(0, 0.9 * self.height(), self.width(),
                          0.1 * self.height())
            rect2 = QRect(0, 0, self.width(), 0.9 * self.height())
            painter.fillRect(rect2, Qt.black)
            eye_find = 0
            if math.isnan(self.p_x[0]) or math.isnan(self.p_x[1]):
                painter.fillRect(rect1, Qt.red)
            else:
                relative_x = 1 - self.p_x[0]
                relative_y = self.p_x[1]
                if relative_x <= 1 and relative_y <= 1:
                    painter.drawEllipse(
                        QPoint(relative_x * self.W, relative_y * self.H),
                        self.current_rad, self.current_rad)
                    eye_find = eye_find + 1
                else:
                    painter.fillRect(rect1, Qt.red)
            if math.isnan(self.p_y[0]) or math.isnan(self.p_y[1]):
                painter.fillRect(rect1, Qt.red)
            else:
                relative_x = 1 - self.p_y[0]
                relative_y = self.p_y[1]
                if relative_x <= 1 and relative_y <= 1:
                    painter.drawEllipse(
                        QPoint(relative_x * self.W, relative_y * self.H),
                        self.current_rad, self.current_rad)
                    eye_find = eye_find + 1
                else:
                    painter.fillRect(rect1, Qt.red)
            if eye_find == 2:
                painter.fillRect(rect1, Qt.green)
            painter.end()


class StartPanel(QMainWindow):
    # custom signals
    start_calibration = Signal()
    begin_test = Signal(str)
    continue_test = Signal(str)

    def __init__(self, parent=None):
        super().__init__(parent)
        self.ui = Ui_start_panel()
        self.ui.setupUi(self)
        self.setWindowTitle(u'实验主控制台')
        self.ui.btn_calibration.setEnabled(False)
        self.ui.btn_start.setEnabled(False)
        self.eye_show = EyePosShow(self.ui.widget_eyepos)
        self.eye_show.setObjectName("eye_show")
        self.ui.widget_eyepos_layout.addWidget(self.eye_show)

        # init some params
        self.is_calibrated = False
        self.eyetracker_frequency = global_config.get_value('eyetracker', 'frequency')
        self.is_debug = global_config.get_value('mode', 'debug')
        self.dir_imgdb = global_config.get_value('database','path')
        self.ui.lineEdit_imgdb_dir.setText(self.dir_imgdb)
        self.participant_id = 'debug'

        self.eyetracker_wrap = global_config.eyetracker_wrapper

        self.calibration_widget = CalibrationWidget()
        self.calibration_result = CalibrationResultWidget()
        self.setting_dialog = SettingDialog()
        self.image_show_widget = ImageShowWidget()

        self.timer = QTimer()
        self.timer.stop()
        self.timer.setInterval(40)
        self.timer.timeout.connect(self.do_timer_timeout)
        
        self.experiment_started = False
        self.eyetracker_subscribed = False
        self.init_connections()

    def init_connections(self):
        self.timer.timeout.connect(
            self.do_timer_timeout)
        self.start_calibration.connect(
            self.calibration_widget.start_calibration)
        self.calibration_widget.calibration_finish.connect(
            self.calibration_result.draw_calibration_samples)
        self.calibration_widget.calibration_finish.connect(
            self.solve_calibration_end)
        self.begin_test.connect(
            self.image_show_widget.begin_test)
        self.continue_test.connect(
            self.image_show_widget.continue_test)
        self.image_show_widget.eye_detection_error.connect(
            self.solve_eye_detection_error)
        self.image_show_widget.experiment_pause.connect(
            self.image_show_pause)
        self.image_show_widget.experiment_finished.connect(
            self.finish_experiment)

    # custom slots
    # slots autoconnect by name
    @Slot()
    def on_action_setting_triggered(self):
        self.eyetracker_wrap.unsubscribe_user_position()
        self.setting_dialog.show()

    @Slot()
    def on_btn_start_eyetracker_clicked(self):
        if not self.eyetracker_wrap.eyetracker:
            self.ui.eyetracker_info.clear()
            self.ui.eyetracker_info.appendPlainText(
                'there is no eyetracker detected, please check your connection and'
                ' settings, then click the button again!')
        else:
            self.eyetracker_wrap.set_frequency(global_config.config_params['eyetracker']['frequency'])
            self.ui.eyetracker_info.clear()
            self.ui.eyetracker_info.appendPlainText(
                'Address:' + self.eyetracker_wrap.address)
            self.ui.eyetracker_info.appendPlainText(
                "Model: " + self.eyetracker_wrap.model)
            self.ui.eyetracker_info.appendPlainText(
                "Name: " + self.eyetracker_wrap.device_name)
            self.ui.eyetracker_info.appendPlainText(
                "Serial number: " + self.eyetracker_wrap.serial_number)
            self.ui.eyetracker_info.appendPlainText(
                "Gaze output frequency: " + str(self.eyetracker_wrap.get_frequency()))
            if not self.eyetracker_subscribed:
                self.eyetracker_wrap.subscribe_user_position()
                self.eyetracker_subscribed = True
            else:
                QMessageBox.information(self,'提示','重新向眼动仪请求位置数据，请稍等1～2秒')
                self.eyetracker_wrap.unsubscribe_user_position()
                self.eyetracker_wrap.subscribe_user_position()
            self.eye_show.IsPainter = True
            self.timer.start()
            self.ui.btn_calibration.setEnabled(True)

    @Slot()
    def on_btn_calibration_clicked(self):
        # query whether to enter calibration
        query_result = QMessageBox.question(self, '提示', '是否开始校准？',
                                            QMessageBox.Yes | QMessageBox.No,
                                            QMessageBox.Yes)
        if (query_result == QMessageBox.Yes):
            self.timer.stop()
            self.start_calibration.emit()
            if self.is_calibrated == False:
                self.is_calibrated = True

    @Slot()
    def on_btn_getdir_clicked(self):
        str_dir = QFileDialog.getExistingDirectory()
        if str_dir == '':
            return
        self.ui.lineEdit_imgdb_dir.setText(str_dir)

    @Slot()
    def on_btn_imgdb_apply_clicked(self):
        self.dir_imgdb = self.ui.lineEdit_imgdb_dir.text()
        is_exist = os.path.exists(self.dir_imgdb)
        if not is_exist:
            QMessageBox.warning(self, u'错误警告', u'路径不存在！')
            return
        global_config.set_value('database', 'path', self.dir_imgdb)
        self.ui.btn_start.setEnabled(True)

    @Slot()
    def on_btn_start_clicked(self):
        if not os.path.exists(self.dir_imgdb):
            msg = '数据库地址：%s 无效,请检查是否已应用' % (self.dir_imgdb)
            QMessageBox.warning(self, '警告', msg)
            return
        if self.eyetracker_wrap.eyetracker == None:
            if (self.is_debug):
                QMessageBox.warning(self, '调试模式', '未检测到眼动仪设备，下面将只显示页面，不记录信息')
            else:
                QMessageBox.warning(self, 'fatal error', '未检测到眼动仪设备，请检查连接')
                return
        if self.experiment_started == False:
            self.experiment_started = True
            self.timer.stop()
            self.begin_test.emit(self.participant_id)
        else:
            self.image_show_widget.is_ready = True
            self.timer.stop()
            self.continue_test.emit(self.participant_id)

    # slots connect manually
    @Slot(str)
    def begin_setting(self, participant_id='debug'):
        self.participant_id = participant_id
        self.show()

    @Slot()
    def do_timer_timeout(self):
        user_position = self.eyetracker_wrap.user_position
        if len(user_position):
            left_data = user_position['left_user_position']
            right_data = user_position['right_user_position']
            self.eye_show.p_x = left_data
            self.eye_show.p_y = right_data
            self.eye_show.update()
            track_box = self.eyetracker_wrap.get_track_box()
            track_box_base = track_box.front_upper_right
            track_box_size = (
                track_box.back_lower_left[0] - track_box.front_upper_right[0],
                track_box.back_lower_left[1] - track_box.front_upper_right[1],
                track_box.back_lower_left[2] - track_box.front_upper_right[2],
            )
            left_z = 0
            right_z = 0
            left_d2c = 1
            right_d2c = 1
            if user_position['left_user_position_validity']:
                left_z = left_data[2] * track_box_size[2] + track_box_base[2]
                left_d2c = math.sqrt((left_data[0] - 0.5)**2 +
                                     (left_data[1] - 0.5)**2 +
                                     (left_data[2] - 0.5)**2)
            if user_position['right_user_position_validity']:
                right_z = right_data[2] * track_box_size[2] + track_box_base[2]
                right_d2c = math.sqrt((right_data[0] - 0.5)**2 +
                                      (right_data[1] - 0.5)**2 +
                                      (right_data[2] - 0.5)**2)
            distance = int(0.5 * (left_z + right_z) / 10)
            d2c = 100 - int(0.5 * (left_d2c + right_d2c) * 100)
            self.ui.pgb_h.setValue(d2c)
            if distance < 45:
                self.ui.pgb_v.setValue(45)
                self.ui.distance.setText(str(45))
            elif distance > 75:
                self.ui.pgb_v.setValue(75)
                self.ui.distance.setText(str(75))
            else:
                self.ui.pgb_v.setValue(distance)
                self.ui.distance.setText(str(distance))

    @Slot()
    def solve_calibration_end(self):
        self.timer.start()
    
    @Slot()
    def solve_eye_detection_error(self):
        self.timer.start()

    @Slot()
    def finish_experiment(self):
        self.eyetracker_wrap.unsubscribe_user_position()
        self.close()

    @Slot()
    def image_show_pause(self):
        self.timer.start()
        self.show()

    # other methods


if __name__ == "__main__":
    global_config.init()
    app = QApplication([])
    widget = StartPanel()
    widget.begin_setting()
    sys.exit(app.exec_())

File: cpp/src/startpanel.cpp
#include "startpanel.h"
#include "ui_startpanel.h"
#include <QDir>
#include <QFileDialog>
#include <QMessageBox>
#include <QtMath>

static TobiiResearchUserPositionGuide global_user_position;
void user_position_guide_callback(TobiiResearchUserPositionGuide *user_position_guide, void *user_data)
{
    memcpy(user_data, user_position_guide, sizeof(*user_position_guide));
};
StartPanel::EyePosShow::EyePosShow(QWidget *parent) : QWidget(parent)
{
    this->eye_pos_adcs[0] = {0, 0, 0}; // left_eye_pos
    this->eye_pos_adcs[1] = {0, 0, 0}; // right_eye_pos
    this->current_rad = 10;
    this->IsPainter = false;
};

void StartPanel::EyePosShow::paintEvent(QPaintEvent *event)
{
    QWidget::paintEvent(event);
    if (this->IsPainter == true)
    {
        this->W = this->width();
        this->H = this->height() * 0.8;
        QPainter painter;
        painter.begin(this);
        {
            QPen pen = QPen(Qt::white, 1, Qt::SolidLine);
            painter.setPen(pen);
            QBrush brush = QBrush(Qt::white, Qt::SolidPattern);
            painter.setBrush(brush);
            QRect rect_status = QRect(0, 0.9 * this->height(),
                                      this->width(), 0.1 * this->height());
            QRect rect_pos = QRect(0, 0,
                                   this->width(), 0.9 * this->height());
            painter.fillRect(rect_pos, Qt::black);
            // paint two eye position
            int eye_find = 0;
            for (auto point_pos : this->eye_pos_adcs)
            {
                if (point_pos[0] > 0 && point_pos[0] < 1 &&
                    point_pos[1] > 0 && point_pos[0] < 1)
                {
                    float relative_x = 1 - point_pos[0];
                    float relative_y = point_pos[1];
                    painter.drawEllipse(
                        QPoint(relative_x * this->W, relative_y * this->H),
                        this->current_rad, this->current_rad);
                    eye_find++;
                }
            }
            // change rect_status color
            if (eye_find == 0)
                painter.fillRect(rect_status, Qt::red);
            else if (eye_find == 1)
                painter.fillRect(rect_status, Qt::yellow);
            else if (eye_find == 2)
                painter.fillRect(rect_status, Qt::green);
        }
        painter.end();
    }
};

StartPanel::StartPanel(QWidget *parent) : QMainWindow(parent),
                                          ui(new Ui::StartPanel)
{
    ui->setupUi(this);
    this->setWindowTitle("实验主控制台");
    this->ui->btn_calibration->setEnabled(false);
    this->ui->btn_start->setEnabled(false);
    this->eye_show = new EyePosShow(ui->widget_eyepos);
    this->eye_show->setObjectName("eye_show");
    this->ui->widget_eyepos_layout->addWidget(this->eye_show);

    // init some params
    this->is_calibrated = false;
    this->eyetracker_frequency = global_config.get_value("eyetracker/frequency", 60.0f).toFloat();
    this->is_debug = global_config.get_value("mode/debug", true).toBool();
    this->dir_imgdb = global_config.get_value("database/path", "./imgdb").toString();
    this->ui->lineEdit_imgdb_dir->setText(this->dir_imgdb);
    this->participant_id = "debug";
    this->eyetracker_wrap = global_config.get_eyetracker_wrapper();
    this->experiment_started = false;
    this->eyetracker_subscribed = false;
    this->timer = new QTimer();
    this->timer->stop();
    this->timer->setInterval(40);
    this->init_connections();
};

StartPanel::~StartPanel()
{
    delete ui;
    delete timer;
};

void StartPanel::init_connections()
{
    connect(this->timer, SIGNAL(timeout()), this, SLOT(do_timer_timeout()));
};

void StartPanel::on_action_setting_triggered()
{
    this->eyetracker_wrap->unsubscribe_user_position(user_position_guide_callback);
    this->eyetracker_subscribed = false;
    this->setting_dialog = new SettingDialog(this);
    connect(this->setting_dialog,SIGNAL(settings_changed()),this,SLOT(on_btn_start_eyetracker_clicked()));
    this->setting_dialog->show();
};

void StartPanel::on_btn_start_eyetracker_clicked()
{
    if (!this->eyetracker_wrap->eyetracker)
    {
        global_config.load_eyetracker();
        this->ui->eyetracker_info->clear();
        this->ui->eyetracker_info->appendPlainText(
            "there is no eyetracker detected, please check your connection and"
            " settings, then click the button again!");
    }
    else
    {
        if (this->eyetracker_subscribed)
        {
            this->timer->stop();
            QMessageBox::information(this, "提示", "重新向眼动仪请求位置数据，请稍等1～2秒");
            this->eyetracker_wrap->unsubscribe_user_position(user_position_guide_callback);
            this->eyetracker_subscribed = false;
        }
        this->ui->eyetracker_info->clear();
        auto eyetracker_info = EyeTrackerWrapper::get_eyetracker_info(this->eyetracker_wrap->eyetracker);
        float frequency = this->eyetracker_wrap->get_current_frequency();
        this->ui->eyetracker_info->appendPlainText(
            "Address:" + eyetracker_info["address"].toString());
        this->ui->eyetracker_info->appendPlainText(
            "Model: " + eyetracker_info["model"].toString());
        this->ui->eyetracker_info->appendPlainText(
            "Name: " + eyetracker_info["device_name"].toString());
        this->ui->eyetracker_info->appendPlainText(
            "Serial number: " + eyetracker_info["serial_number"].toString());
        this->ui->eyetracker_info->appendPlainText(
            QString("Gaze output frequency: %1").arg(frequency));
        this->eyetracker_wrap->subscribe_user_position(user_position_guide_callback,global_user_position);
        this->eyetracker_subscribed = true;
        this->eye_show->IsPainter = true;

        this->timer->start();
        this->ui->btn_calibration->setEnabled(true);
    }
};

void StartPanel::on_btn_calibration_clicked()
{
    // query whether to enter calibration
    auto query_result = QMessageBox::question(this, "提示", "是否开始校准？",
                                              QMessageBox::Yes | QMessageBox::No,
                                              QMessageBox::Yes);
    if (query_result == QMessageBox::Yes)
    {
        this->timer->stop();
        this->calibration_widget = new CalibrationWidget();
        this->calibration_result = new CalibrationResultWidget();
        connect(this, SIGNAL(start_calibration()), this->calibration_widget, SLOT(start_calibration()));
        connect(this->calibration_widget, SIGNAL(calibration_finish(TobiiResearchCalibrationResult*)),
                this->calibration_result, SLOT(draw_calibration_samples(TobiiResearchCalibrationResult*)));
        connect(this->calibration_widget, SIGNAL(calibration_finish(TobiiResearchCalibrationResult*)),
                this, SLOT(solve_calibration_end()));
        emit start_calibration();
        if(this->is_debug){
            this->is_calibrated = true;
        }

    }
};

void StartPanel::on_btn_getdir_clicked()
{
    QString str_dir = QFileDialog::getExistingDirectory();
    if (str_dir.trimmed().size() <= 0)
        return;
    this->ui->lineEdit_imgdb_dir->setText(str_dir);
};

void StartPanel::on_btn_imgdb_apply_clicked()
{
    this->dir_imgdb = this->ui->lineEdit_imgdb_dir->text();
    QDir dir;
    bool is_exist = dir.exists(this->dir_imgdb);
    if (!is_exist)
    {
        QMessageBox::warning(this, "错误警告", "路径不存在！");
        return;
    }
    global_config.set_value("database/path", this->dir_imgdb);
    if(this->is_calibrated)
        this->ui->btn_start->setEnabled(true);
    else if(this->is_debug){
        QMessageBox::warning(this,"调试模式","眼动仪未矫正!");
        this->ui->btn_start->setEnabled(true);
    }
};

void StartPanel::on_btn_start_clicked()
{
    QDir dir(this->dir_imgdb);
    if (!dir.exists())
    {
        QString msg = QString("数据库地址：%1 无效,请检查是否已应用").arg(this->dir_imgdb);
        QMessageBox::warning(this, "警告", msg);
        return;
    }
    if (!this->eyetracker_wrap->eyetracker){
        if (this->is_debug)
            QMessageBox::warning(this, "调试模式", "未检测到眼动仪设备，下面将只显示页面，不记录信息");
        else
        {
            QMessageBox::warning(this, "错误", "未检测到眼动仪设备，请检查连接");
            return;
        }
    }
    if (!this->experiment_started)
    {
        this->experiment_started = true;
        this->timer->stop();
        this->image_show_widget = new ImageShowWidget();
        connect(this, SIGNAL(begin_test(QString)), this->image_show_widget, SLOT(begin_test(QString)));
        connect(this, SIGNAL(continue_test(QString)), this->image_show_widget, SLOT(continue_test(QString)));
        connect(this->image_show_widget, SIGNAL(eye_detection_error(QString)), this, SLOT(solve_eye_detection_error()));
        connect(this->image_show_widget, SIGNAL(experiment_pause(QString)), this, SLOT(image_show_pause()));
        connect(this->image_show_widget, SIGNAL(experiment_finished(QString)), this, SLOT(finish_experiment()));
        emit begin_test(this->participant_id);
    }
    else
    {
        this->timer->stop();
        emit continue_test(this->participant_id);
    }
};

void StartPanel::begin_setting(QString participant_id)
{
    this->participant_id = participant_id;
    this->show();
};

void StartPanel::do_timer_timeout()
{
    TobiiResearchUserPositionGuide &user_position = global_user_position;
    if (user_position.left_eye.validity || user_position.right_eye.validity)
    {
        MyFPoint3D left_data = {
            user_position.left_eye.user_position.x,
            user_position.left_eye.user_position.y,
            user_position.left_eye.user_position.z};
        MyFPoint3D right_data = {
            user_position.right_eye.user_position.x,
            user_position.right_eye.user_position.y,
            user_position.right_eye.user_position.z};

        this->eye_show->eye_pos_adcs[0] = left_data;
        this->eye_show->eye_pos_adcs[1] = right_data;
        this->eye_show->update();
        TobiiResearchTrackBox track_box = this->eyetracker_wrap->get_track_box();
        MyFPoint3D track_box_base = {
            track_box.front_upper_right.x,
            track_box.front_upper_right.y,
            track_box.front_upper_right.z,
        };
        MyFPoint3D track_box_size = {track_box.back_lower_left.x - track_box.front_upper_right.x,
                                     track_box.back_lower_left.y - track_box.front_upper_right.y,
                                     track_box.back_lower_left.z - track_box.front_upper_right.z};
        float left_z = 0, right_z = 0, left_d2c = 1, right_d2c = 1;
        if (user_position.left_eye.validity)
        {
            left_z = left_data[2] * track_box_size[2] + track_box_base[2];
            left_d2c = qSqrt(qPow((left_data[0] - 0.5), 2) +
                             qPow((left_data[1] - 0.5), 2) +
                             qPow((left_data[2] - 0.5), 2));
        }
        if (user_position.right_eye.validity)
        {
            right_z = right_data[2] * track_box_size[2] + track_box_base[2];
            right_d2c = qSqrt(qPow((right_data[0] - 0.5), 2) +
                              qPow((right_data[1] - 0.5), 2) +
                              qPow((right_data[2] - 0.5), 2));
        }
        int distance = 0.5 * (left_z + right_z) / 10;
        int d2c = 100 - int(0.5 * (left_d2c + right_d2c) * 100);
        this->ui->pgb_h->setValue(d2c);
        if (distance < 45)
        {
            this->ui->pgb_v->setValue(45);
            this->ui->distance->setText("45");
        }
        else if (distance > 75)
        {
            this->ui->pgb_v->setValue(75);
            this->ui->distance->setText("75");
        }
        else
        {
            this->ui->pgb_v->setValue(distance);
            this->ui->distance->setText(QString("%1").arg(distance));
        }
    }else{
        this->eye_show->eye_pos_adcs={};
        this->eye_show->update();
        this->ui->pgb_h->setValue(0);
    }
};

void StartPanel::solve_calibration_end()
{
    if (!this->is_calibrated)
        this->is_calibrated = true;
    QDir dir(this->dir_imgdb);
    if(dir.exists()){
        this->ui->btn_start->setEnabled(true);
    };
    this->timer->start();
};

void StartPanel::solve_eye_detection_error()
{
    this->ui->btn_start->setText("继续实验");
    this->timer->start();
};

void StartPanel::finish_experiment()
{
    this->eyetracker_wrap->unsubscribe_user_position(&user_position_guide_callback);
    QMessageBox::information(this,"实验结束","感谢您的参与!");
    this->image_show_widget->close();
    this->close();
};

void StartPanel::image_show_pause()
{
    this->timer->start();
    this->show();
};
